### PR TITLE
Feature/update harbor cluster interface

### PR DESCRIPTION
**Pull request by wangcanfengxs**
CHANGES:
- add harbor cluster main reconciler
- every components implements lcm.Controller interface 
- remove service.ServiceReconciler interface

### CODE REVIEW

File: controllers/goharbor/harborcluster/cache/redis.go [hunk 1]
```diff
@@ -1,13 +1,21 @@
 package cache
 
 import (
-	goharborv1alpha2 "github.com/goharbor/harbor-operator/apis/goharbor.io/v1alpha2"
+	"github.com/goharbor/harbor-operator/apis/goharbor.io/v1alpha2"
 	"github.com/goharbor/harbor-operator/pkg/lcm"
 )
 
 type RedisReconciler struct {
```
Comment (steven-zou) [inline controllers/goharbor/harborcluster/cache/redis.go:8]: It seems we don't need such a cache reconciler here. The harborcluster reconcile should call the cache lcm controller to do the harborcluster reconcile process.

File: controllers/goharbor/harborcluster/database/postgresql.go [hunk 2]
```diff
@@ -1,13 +1,21 @@
 package database
 
 import (
-	goharborv1alpha2 "github.com/goharbor/harbor-operator/apis/goharbor.io/v1alpha2"
+	"github.com/goharbor/harbor-operator/apis/goharbor.io/v1alpha2"
 	"github.com/goharbor/harbor-operator/pkg/lcm"
 )
 
 type PostgreSQLReconciler struct {
```
Comment (steven-zou) [inline controllers/goharbor/harborcluster/database/postgresql.go:8]: see above

File: controllers/goharbor/harborcluster/harbor/harbor.go [hunk 3]
```diff
@@ -8,7 +8,14 @@ import (
 type HarborReconciler struct {
```
Comment (steven-zou) [inline controllers/goharbor/harborcluster/harbor/harbor.go:8]: see above comment

File: controllers/goharbor/harborcluster/service.go [hunk 4]
```diff
@@ -14,23 +13,18 @@ import (
 	"k8s.io/client-go/tools/record"
 )
 
-type ServiceReconciler interface {
-	// Reconcile the dependent service.
-	Reconcile(harborCluster *v1alpha2.HarborCluster) (*lcm.CRStatus, error)
-}
-
 type ServiceGetter interface {
```
Comment (steven-zou) [inline controllers/goharbor/harborcluster/service.go:16]: I think the lcm controllers can be created in the `New` method of the HarborCluster Reconciler struct.

File: controllers/goharbor/harborcluster/storage/minio.go [hunk 5]
```diff
@@ -1,13 +1,21 @@
 package storage
 
 import (
-	goharborv1alpha2 "github.com/goharbor/harbor-operator/apis/goharbor.io/v1alpha2"
+	"github.com/goharbor/harbor-operator/apis/goharbor.io/v1alpha2"
 	"github.com/goharbor/harbor-operator/pkg/lcm"
 )
 
 type MinIOReconciler struct {
```
Comment (steven-zou) [inline controllers/goharbor/harborcluster/storage/minio.go:8]: Same to the above comment

**Review by steven-zou — COMMENTED** [review]
commented

File: controllers/goharbor/harborcluster/service.go (hunk 4 above)
Comment (wangcanfengxs) [reply controllers/goharbor/harborcluster/service.go:16]: ServiceGetter is created in New method

File: controllers/goharbor/harborcluster/cache/redis.go [hunk 6]
```diff
@@ -1,13 +1,25 @@
...
+	"github.com/goharbor/harbor-operator/apis/goharbor.io/v1alpha2"
 	"github.com/goharbor/harbor-operator/pkg/lcm"
 )
 
-type RedisReconciler struct {
+type RedisController struct {
 }
 
-func (r RedisReconciler) Reconcile(harborCluster *goharborv1alpha2.HarborCluster) (*lcm.CRStatus, error) {
+func (r *RedisController) Apply(harborcluster *v1alpha2.HarborCluster) (*lcm.CRStatus, error) {
```
Comment (soulseen) [inline controllers/goharbor/harborcluster/cache/redis.go:11]: how about the params `harborcluster` moved to struct `MinIOController`? and set it by `NewCacheController`.

**Review by Ghostbaby — APPROVED** [review]
lgtm

**Review by soulseen — APPROVED** [review]
lgtm

**Pull request merged**